### PR TITLE
API Aggregator

### PR DESCRIPTION
**Pull request by chenopis**
[kubernetes/features#263](https://github.com/kubernetes/features/issues/263)

/assign @cheftako

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/4173)

### CODE REVIEW

**Comment by chenopis** [comment]
@lavalamp This is just the initial commit. I will fold in the changes from the Google doc once you guys have that ironed out.

**Comment by chenopis** [comment]
@abiogenesis-now Can you help w/ a docs review? Thx

File: docs/concepts/architecture/apiserver-aggregation.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+In 1.7, the Aggregator runs in a process with the kube-apiserver; however, until an extension resource is registered, the Aggregator will do nothing. Once a new resource is registered, Aggregator will proxy the request to the extension-apiserver to be served. 
+
+**Note:** The extension-apiserver will normally need to be paired with a controller to actually provide any functionality. As a result, things like the apiserver-builder will actually provide a skeleton for both. The service-catalog should provide both the extension-apiserver and controller for the services it provides.
+
+{% endcapture %}
+
+{% capture whatsnext %}
+
+* To get the aggregator working in your environment, see [Setup the API aggregator](/docs/tasks/access-kubernetes-api/setup-api-aggregator/).
+* Learn how to [Extend the Kubernetes API Using Third Party Resources](/docs/tasks/access-kubernetes-api/extend-api-third-party-resource/).
```
Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: Looks like TPRs are being deprecated in 1.7? Swap out with [Extend the Kubernetes API Using Custom Resource Definitions](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/).

File: docs/concepts/architecture/apiserver-aggregation.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+---
+title: Extending the Kubernetes API with Aggregator
+assignees:
+- chenopis
+---
+
+{% capture overview %}
+
+The *Aggregator* is a system to allow the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs.
```
Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: Is it best practice to refer to the "apiserver" or "API Server"? I've seen it both ways---it's more of a nit, but consistency would be nice.

Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: I think the sentence reads a little easier as: "The *Aggregator* is a system that allows the Kubernetes apiserver to be extended with additional APIs, beyond what is offered by the core Kubernetes APIs."?

Let me know if this comment is too fine-grained.

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 1 above)
Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: Also, [this](https://docs.google.com/document/d/1lU1SnVtEec2iIfYx5U3L5N0za2YhfEOik0uPen276Ks/edit#heading=h.6ae9qgk4mhne) might be the Google doc that you are already aware of, but it seems like TPR/CRDs are distinct from API Aggregation? It would be great to make this more clear, since this is a high-level conceptual doc.

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+1. Create a Kubernetes cluster role for the operations you want to allow on your resources.
+1. Create a Kubernetes cluster role binding from the default service account in your namespace to the cluster role you just created.
+1. Create a Kubernetes apiservice. The CA cert above should be base 64 encoded, stripped of new lines and used as the spec.caBundle in the apiservce.  This should not be namespaced.
+1. Use kubectl to get your resource. It will probably return "No resources found." Which means that everything worked but you currently have no objects of that resource type created yet.
+
+{% endcapture %}
+
+{% capture whatsnext %}
+
+* For more information, see the [Aggregator Networking Requirements](https://docs.google.com/document/d/1KNT4iS_Y2miLARrfSPumBIiFo_h7eb5B2pVOZJ0ZmjQ/edit) doc.
```
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: I don't have access to this Google doc---might want to check the permissions on it.

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+* For a high level overview, see [Extending the Kubernetes API with Aggregator](/docs/concepts/architecture/apiserver-aggregation/).
+* Learn how to [Extend the Kubernetes API Using Third Party Resources](/docs/tasks/access-Kubernetes-api/extend-api-third-party-resource/).
```
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: see the earlier comment about CRDs

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
+---
+title: Setup the API Aggregator
+assignees:
+- chenopis
+---
+
+{% capture overview %}
+
+Setting up the API Aggregator will allow the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs.
```
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: nit: "allows" instead of "will allow" (there are a couple unnecessary future tenses scattered around...might want to just Ctrl-f "will")

File: docs/concepts/architecture/apiserver-aggregation.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+{% endcapture %}
+
+{% capture body %}
+
+## Overview
+
+The Aggregator allows both third party extension via something like [service-catalog](https://github.com/kubernetes-incubator/service-catalog/blob/master/README.md) and the creation of extensions to allow seamless addition of the resources a customer needs. The latter can be easily bootstrapped using [apiserver-builder](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/README.md). 
+
+In 1.7, the Aggregator runs in a process with the kube-apiserver; however, until an extension resource is registered, the Aggregator will do nothing. Once a new resource is registered, Aggregator will proxy the request to the extension-apiserver to be served.
```
Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: nit: "to be served" might be redundant

File: docs/concepts/architecture/apiserver-aggregation.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+{% capture body %}
+
+## Overview
+
+The Aggregator allows both third party extension via something like [service-catalog](https://github.com/kubernetes-incubator/service-catalog/blob/master/README.md) and the creation of extensions to allow seamless addition of the resources a customer needs. The latter can be easily bootstrapped using [apiserver-builder](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/README.md). 
+
+In 1.7, the Aggregator runs in a process with the kube-apiserver; however, until an extension resource is registered, the Aggregator will do nothing. Once a new resource is registered, Aggregator will proxy the request to the extension-apiserver to be served. 
+
+**Note:** The extension-apiserver will normally need to be paired with a controller to actually provide any functionality. As a result, things like the apiserver-builder will actually provide a skeleton for both. The service-catalog should provide both the extension-apiserver and controller for the services it provides.
```
Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: ^ a bit wordy----am I correctly interpreting that _both_ the apiserver-builder and the service-catalog provide the extension-apiserver and the controller? (A little confused if "should provide" means that service-catalog does it by default)

File: docs/concepts/architecture/apiserver-aggregation.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+The *Aggregator* is a system to allow the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs. 
+
+{% endcapture %}
+
+{% capture body %}
+
+## Overview
+
+The Aggregator allows both third party extension via something like [service-catalog](https://github.com/kubernetes-incubator/service-catalog/blob/master/README.md) and the creation of extensions to allow seamless addition of the resources a customer needs. The latter can be easily bootstrapped using [apiserver-builder](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/README.md).
```
Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: It's not entirely clear to me what the difference between these two are, sorry! I'll read up on it later and see if I have suggestions, but do you mind clarifying?

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 6 above)
Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: also, is *extension-apiserver* an official term? I can infer what you're referring to, but it definitely made me pause. Would be nice to italicize it and mention it in the first paragraph.

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+Setting up the API Aggregator will allow the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs. 
+
+{% endcapture %}
+
+{% capture prerequisites %}
+
+* You need to have a Kubernetes cluster running.
+* Make sure there are Aggregator certificates (both CA and proxy) setup on the Kubernetes apiserver.
+* The proxy cert must have been signed by the relevant CA.
```
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: clarify what "relevant" means?

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    --proxy-client-key-file=<path to aggregator proxy key>
+
+If you are not running kube-proxy on your master, as the [Kubernetes Architectural Roadmap](https://docs.google.com/a/google.com/document/d/1XkjVm4bOeiVkj-Xt1LgoGiqWsBfNozJ51dyI-ljzt1o/edit?usp=sharing) recommends, then you need to make sure that the system is enabled with the following apiserver flag. Again, this may have already been taken care of by your provider.
+
+    --enable-aggregator-routing=true
+
+## Get an extension api-server working with Aggregator
+
+The generated skeleton, which comes with [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), should setup the skeleton server it creates to do all of these steps. You can also look at the sample-apiserver under Kubernetes for ways the following can be done.
+
```
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: provide link to sample-apiserver? Also consider rewording the first sentence.

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+1. Create the Kubernetes namespace you want to run your extension api-service in.
+1. Create/get a CA cert to be used to sign the server cert the extension api-server uses for HTTPS.
```
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: do these easily map to code? sample commands might make it easier to follow.

**Comment by chenopis** [comment]
@cheftako Can you suggest someone to do the tech review? Thx

**Comment by cheftako** [comment]
I'm starting the review now.

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 8 above)
Comment (cheftako) [inline docs/concepts/architecture/apiserver-aggregation.md]: Third party extensions are resource which are not built by either the cluster owner or the kubernetes team. Service Catalog is an example of this and is in fact an example which will make it easy for cluster owners to easily add other third party extensions to their cluster. This sort of extension is already built but does need to be deployed properly to the kubernetes cluster. If the cluster owner wants to create their own extension they first have to build something which adheres to the extension api-server requirements and then they have to work out how to deploy it correctly. So the story difference here is largely about the use case, though the technical details are very similar.

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 6 above)
Comment (cheftako) [inline docs/concepts/architecture/apiserver-aggregation.md]: Honestly I think we're still trying to come up with the official term. However I've heard several complaints about user-apiserver as that seems to exclude the third party use case. As this is meant as an extension mechanism we seem to be moving toward consensus on that term.

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 7 above)
Comment (cheftako) [inline docs/concepts/architecture/apiserver-aggregation.md]: Yes

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 9 above)
Comment (cheftako) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: Kubernetes (and the kube-apiserver) have multiple CAs. I want to make sure the proxy is signed by the aggregator CA and not say the master CA.

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 10 above)
Comment (cheftako) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: https://github.com/kubernetes/sample-apiserver/blob/master/README.md

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+    --requestheader-client-ca-file=<path to aggregator CA cert>
+    --requestheader-allowed-names=aggregator
+    --requestheader-extra-headers-prefix=X-Remote-Extra-
+    --requestheader-group-headers=X-Remote-Group
+    --requestheader-username-headers=X-Remote-User
+    --proxy-client-cert-file=<path to aggregator proxy cert>
+    --proxy-client-key-file=<path to aggregator proxy key>
+
+If you are not running kube-proxy on your master, as the [Kubernetes Architectural Roadmap](https://docs.google.com/a/google.com/document/d/1XkjVm4bOeiVkj-Xt1LgoGiqWsBfNozJ51dyI-ljzt1o/edit?usp=sharing) recommends, then you need to make sure that the system is enabled with the following apiserver flag. Again, this may have already been taken care of by your provider.
```
Comment (cheftako) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: I found this a tad confusing. Just to be clear the roadmap recommends not running kube-proxy on the master.

**Comment by chenopis** [comment]
@cheftako Can we also get a another set of eyes on this for a second tech review?

**Comment by cheftako** [comment]
@lavalamp has agreed to take a look.

File: _data/concepts.yml [hunk 13]
```diff
@@ -20,6 +20,7 @@ toc:
   section:
   - docs/concepts/architecture/nodes.md
   - docs/concepts/architecture/master-node-communication.md
+  - docs/concepts/architecture/apiserver-aggregation.md
```
Comment (lavalamp) [inline _data/concepts.yml]: I agree this should probably go along with the api section, or we could make a new api extension section and include this, TPR, CRD, & admission controllers / initializers.

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 2 above)
Comment (lavalamp) [inline docs/concepts/architecture/apiserver-aggregation.md]: I would like to refer to this as the "aggregation layer" since it is built in and not a separate component.

Comment (lavalamp) [inline docs/concepts/architecture/apiserver-aggregation.md]: "The aggregation layer allows Kubernetes to be extended with additional APIs which are not included in the kubernetes core APIs"

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 8 above)
Comment (lavalamp) [inline docs/concepts/architecture/apiserver-aggregation.md]: "aggregation layer"

Comment (lavalamp) [inline docs/concepts/architecture/apiserver-aggregation.md]: "The aggregation layer allows installing additional Kubernetes-style APIs in your cluster. These can either be pre-built, existing 3rd party solutions (e.g., service catalog), or user-created APIs (apiserver-builder can get you started)."

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 6 above)
Comment (lavalamp) [inline docs/concepts/architecture/apiserver-aggregation.md]: My suggested wording in the doc was:

"In 1.7 the aggregation layer runs in-process with the kube-apiserver. Until an extension resource is registered, the aggregation layer will do nothing. Users can add a APIService object to register their API, which “claims” the URL path in the Kubernetes API. At this point, the aggregation layer will proxy anything sent to that api path (e.g. /apis/myextension.mycompany.io/v1/…) to the registered APIService. Ordinarily, the APIService will be implemented by an extension-apiserver in a pod running in the cluster. This extension-apiserver will normally need to be paired with one or more controllers if active management of the added resources is needed. As a result, the apiserver-builder will actually provide a skeleton for both. For example, when the service-catalog is installed, it provides both the extension-apiserver and controller for the services it provides."

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 1 above)
Comment (lavalamp) [inline docs/concepts/architecture/apiserver-aggregation.md]: Please don't call these third party resources. Extension APIs?

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 5 above)
Comment (lavalamp) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: I would call this doc "turn on the aggregation layer" or "configure the aggregation layer" or "configure API Aggregation"

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+- chenopis
+---
+
+{% capture overview %}
+
+Setting up the API Aggregator will allow the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs. 
+
+{% endcapture %}
+
+{% capture prerequisites %}
```
Comment (lavalamp) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: These are not prereqs, these are what this doc is going to tell you how to do.

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    --requestheader-group-headers=X-Remote-Group
+    --requestheader-username-headers=X-Remote-User
+    --proxy-client-cert-file=<path to aggregator proxy cert>
+    --proxy-client-key-file=<path to aggregator proxy key>
+
+If you are not running kube-proxy on your master, as the [Kubernetes Architectural Roadmap](https://docs.google.com/a/google.com/document/d/1XkjVm4bOeiVkj-Xt1LgoGiqWsBfNozJ51dyI-ljzt1o/edit?usp=sharing) recommends, then you need to make sure that the system is enabled with the following apiserver flag. Again, this may have already been taken care of by your provider.
+
+    --enable-aggregator-routing=true
+
+## Get an extension api-server working with Aggregator
```
Comment (lavalamp) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: IMO this and everything below should go in a separate doc, "how to setup an extension API server".

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 8 above)
Comment (abiogenesis-now) [inline docs/concepts/architecture/apiserver-aggregation.md]: yes this is much more clear!

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 14 above)
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: Ok. Are there any other prereqs then that they'll need before they can get started?

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 9 above)
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: Seems like this should go together with the previous bullet point then?

**Comment by chenopis** [comment]
@lavalamp Ok, thanks for the feedback. I'll address your feedback this weekend.

File: _data/concepts.yml (hunk 13 above)
Comment (chenopis) [inline _data/concepts.yml]: Ok, I'm going to create a new subsection under Concepts called *Extending the Kubernetes API* and put this in there. We can then start adding other docs into that. I did originally try to put it in the Overview subsection next to *Kubernetes API*, but this doc isn't really an overview so it didn't quite fit.

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 2 above)
Comment (chenopis) [inline docs/concepts/architecture/apiserver-aggregation.md]: Yeah, that makes a lot of sense.

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 8 above)
Comment (chenopis) [inline docs/concepts/architecture/apiserver-aggregation.md]: changed

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 6 above)
Comment (chenopis) [inline docs/concepts/architecture/apiserver-aggregation.md]: swapped in

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 7 above)
Comment (chenopis) [inline docs/concepts/architecture/apiserver-aggregation.md]: I've reworded this paragraph to clarify.

File: docs/concepts/architecture/apiserver-aggregation.md (hunk 1 above)
Comment (chenopis) [inline docs/concepts/architecture/apiserver-aggregation.md]: the link has been changed

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 12 above)
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: I reworded it

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 15 above)
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: Ok, I'm splitting this into two docs.

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 4 above)
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: got it

File: docs/tasks/access-kubernetes-api/setup-api-aggregator.md (hunk 3 above)
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-api-aggregator.md]: I'm going to take this out until it's publicly accessible.

**Comment by chenopis** [comment]
I've addressed all the comments. @cheftako @lavalamp @abiogenesis-now PTAL.

Previews:
Concept/[Extending the Kubernetes API with the aggregation layer](https://deploy-preview-4173--kubernetes-io-master-staging.netlify.com/docs/concepts/api-extension/apiserver-aggregation/)
Task/[Configure the aggregation layer](https://deploy-preview-4173--kubernetes-io-master-staging.netlify.com/docs/tasks/access-kubernetes-api/configure-aggregation-layer/)
Task/[Setup an extension API server](https://deploy-preview-4173--kubernetes-io-master-staging.netlify.com/docs/tasks/access-kubernetes-api/setup-extension-api-server/)

**Comment by chenopis** [comment]
FYI, all feedback must be addressed and LGTMs given by EOD Tue, June 27th so that this can be merged for the 1.7 release on June 28th.

File: docs/concepts/api-extension/apiserver-aggregation.md [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+In 1.7 the aggregation layer runs in-process with the kube-apiserver. Until an extension resource is registered, the aggregation layer will do nothing. Users can add an APIService object to register their API, which “claims” the URL path in the Kubernetes API. At that point, the aggregation layer will proxy anything sent to that API path (e.g. /apis/myextension.mycompany.io/v1/…) to the registered APIService. Ordinarily, the APIService will be implemented by an extension-apiserver in a pod running in the cluster. This extension-apiserver will normally need to be paired with one or more controllers if active management of the added resources is needed. As a result, the apiserver-builder will actually provide a skeleton for both. For example, when the service-catalog is installed, it provides both the extension-apiserver and controller for the services it provides.
```
Comment (abiogenesis-now) [inline docs/concepts/api-extension/apiserver-aggregation.md]: want to rewrite the third sentence:

> To register their API, users must add an APIService object, which "claims" the URL path in the Kubernetes API.

in order to better contrast with the second sentence (the aggregation layer doing nothing unless extension resource is registered)

Comment (abiogenesis-now) [inline docs/concepts/api-extension/apiserver-aggregation.md]: I want to gut check my own understanding of things, so quick question: 

**Is it the case that kube-apiserver is an actual official binary, whereas extension-apiserver is a stand-in name that can refer to *any* roll-your-own custom API server?**

(e.g. I should not expect to be able to literally run `./extension-apiserver` unless I explicitly name the custom API server binary that way?)

File: docs/tasks/access-kubernetes-api/configure-aggregation-layer.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+**Note:** There are a few setup requirements for getting the aggregation layer working in your environment to support mutual TLS auth between the proxy and extension apiservers. Kubernetes and the kube-apiserver have multiple CAs, so make sure that the proxy is signed by the aggregation layer CA and not by something else, like the master CA.
+
+{% endcapture %}
+
+{% capture steps %}
+
+## Enable apiserver flags
+
+Enable the aggregation layer via the following apiserver flags. They may have already been taken care of by your provider.
```
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/configure-aggregation-layer.md]: *Clarification:* Are these flags meant to be applied when I run `kube-apiserver`?  

I'm pretty sure this is the case but I could be wrong. Either way, could we clarify that "apiserver" means `kube-apiserver`, given that there are now multiple API servers floating around?

File: docs/concepts/api-extension/apiserver-aggregation.md (hunk 16 above)
Comment (abiogenesis-now) [inline docs/concepts/api-extension/apiserver-aggregation.md]: is the aggregation layer proxying mentioned here something that `kube-apiserver` does or `kube-proxy`?

File: docs/tasks/access-kubernetes-api/setup-extension-api-server.md [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+* You need to have a Kubernetes cluster running.
+* You must [configure the aggregation layer](/docs/tasks/configure-aggregation-layer) and enable the apiserver flags.
+
+{% endcapture %}
+
+{% capture steps %}
+
+## Setup an extension api-server to work with the aggregation layer
+
+If you use an existing 3rd party solution, such as [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), it should generate a skeleton and perform all of the following steps. Alternatively, you can also look at the [sample-apiserver](https://github.com/kubernetes/sample-apiserver/blob/master/README.md) under Kubernetes for ways the following can be done.
```
Comment (abiogenesis-now) [inline docs/tasks/access-kubernetes-api/setup-extension-api-server.md]: I was a bit confused by this blurb at first (e.g. why the steps below lacked detail). Consider the following wording for more clarity?

> The following steps describe how to set up an extension-apiserver *at a high level*. For a concrete example of how they can be implemented, you can look at the [sample-apiserver](https://github.com/kubernetes/sample-apiserver/blob/master/README.md) in the Kubernetes repo.
>
> Alternatively, you can use an existing 3rd party solution, such as [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), which should generate a skeleton and automate all of the following steps for you.

File: docs/concepts/api-extension/apiserver-aggregation.md [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+**Note:** The extension-apiserver will need to be paired with a controller to provide any functionality. Fortunately, if you use [apiserver-builder](https://github.com/kubernetes-incubator/apiserver-builder/blob/master/README.md), it will provide both a skeleton extension-apiserver and a skeleton controller. Alternatively, the [service-catalog](https://github.com/kubernetes-incubator/service-catalog/blob/master/README.md) can install extensions for you and should provide/deploy both the extension apiserver and any other resources it needs.
```
Comment (abiogenesis-now) [inline docs/concepts/api-extension/apiserver-aggregation.md]: by entire part I mean line 23 😅

Comment (chenopis) [inline docs/concepts/api-extension/apiserver-aggregation.md]: done

File: docs/concepts/api-extension/apiserver-aggregation.md (hunk 16 above)
Comment (chenopis) [inline docs/concepts/api-extension/apiserver-aggregation.md]: @cheftako @lavalamp Can you clarify?

Comment (chenopis) [inline docs/concepts/api-extension/apiserver-aggregation.md]: done

File: docs/tasks/access-kubernetes-api/configure-aggregation-layer.md [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+title: Configure the aggregation layer
+assignees:
+- lavalamp
+- cheftako
+- chenopis
+---
+
+{% capture overview %}
+
+Configuring the aggregation layer allows the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs.
```
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/configure-aggregation-layer.md]: good idea

File: docs/tasks/access-kubernetes-api/configure-aggregation-layer.md (hunk 17 above)
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/configure-aggregation-layer.md]: @cheftako @lavalamp Can you confirm that "apiserver" means `kube-apiserver` here?

File: docs/tasks/access-kubernetes-api/configure-aggregation-layer.md [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+The [Kubernetes Architectural Roadmap](https://docs.google.com/a/google.com/document/d/1XkjVm4bOeiVkj-Xt1LgoGiqWsBfNozJ51dyI-ljzt1o/edit?usp=sharing) recommends not running kube-proxy on the master. If you follow this recommendation, then you must make sure that the system is enabled with the following apiserver flag. Again, this may have already been taken care of by your provider.
+
+    --enable-aggregator-routing=true
+
+{% endcapture %}
+
+{% capture whatsnext %}
+
+* [Setup an extension api-server](/docs/tasks/access-kubernetes-api/setup-extension-api-server/) to work with the aggregation layer.
+* For a high level overview, see [Extending the Kubernetes API with Aggregator](/docs/concepts/api-extension/apiserver-aggregation/).
```
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/configure-aggregation-layer.md]: Good catch. Sorry, was updating things out of order.

File: docs/tasks/access-kubernetes-api/setup-extension-api-server.md [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+{% capture overview %}
+
+Setting up an extension API server to work the aggregation layer allows the Kubernetes apiserver to be extended with additional APIs, which are not part of the core Kubernetes APIs. 
+
+{% endcapture %}
+
+{% capture prerequisites %}
+
+* You need to have a Kubernetes cluster running.
+* You must [configure the aggregation layer](/docs/tasks/configure-aggregation-layer) and enable the apiserver flags.
```
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-extension-api-server.md]: good catch

File: docs/tasks/access-kubernetes-api/setup-extension-api-server.md [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+
+{% endcapture %}
+
+{% capture steps %}
+
+## Setup an extension api-server to work with the aggregation layer
+
+If you use an existing 3rd party solution, such as [apiserver-builder](https://github.com/Kubernetes-incubator/apiserver-builder/blob/master/README.md), it should generate a skeleton and perform all of the following steps. Alternatively, you can also look at the [sample-apiserver](https://github.com/kubernetes/sample-apiserver/blob/master/README.md) under Kubernetes for ways the following can be done.
+
+1. Make sure the APIService API is enabled (check --runtime-config); it is on by default, so it should be on unless it’s been deliberately turned off in your cluster.
```
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-extension-api-server.md]: done

File: docs/tasks/access-kubernetes-api/setup-extension-api-server.md [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+1. Create a Kubernetes cluster role binding from the default service account in your namespace to the cluster role you just created.
+1. Create a Kubernetes apiservice. The CA cert above should be base 64 encoded, stripped of new lines and used as the spec.caBundle in the apiservce.  This should not be namespaced.
+1. Use kubectl to get your resource. It should return "No resources found." Which means that everything worked but you currently have no objects of that resource type created yet.
+
+{% endcapture %}
+
+{% capture whatsnext %}
+
+* If you haven't already, [configure the aggregation layer](/docs/tasks/access-kubernetes-api/configure-aggregation-layer/) and enable the apiserver flags.
+* For a high level overview, see [Extending the Kubernetes API with Aggregator](/docs/concepts/api-extension/apiserver-aggregation/).
```
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-extension-api-server.md]: duh. thanks!

File: docs/concepts/api-extension/apiserver-aggregation.md (hunk 16 above)
Comment (chenopis) [inline docs/concepts/api-extension/apiserver-aggregation.md]: 3rd sentence swapped out

File: docs/tasks/access-kubernetes-api/setup-extension-api-server.md (hunk 18 above)
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/setup-extension-api-server.md]: done

**Comment by chenopis** [comment]
@abiogenesis-now I think I put in all your feedback. Let me know if I missed anything.

**Comment by chenopis** [comment]
@cheftako @lavalamp PTAL

**Comment by abiogenesis-now** [comment]
LGTM---thanks for incorporating feedback!

File: docs/concepts/api-extension/apiserver-aggregation.md [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+
+In 1.7 the aggregation layer runs in-process with the kube-apiserver. Until an extension resource is registered, the aggregation layer will do nothing. To register their API, users must add an APIService object, which "claims" the URL path in the Kubernetes API. At that point, the aggregation layer will proxy anything sent to that API path (e.g. /apis/myextension.mycompany.io/v1/…) to the registered APIService. 
+
+Ordinarily, the APIService will be implemented by an *extension-apiserver* in a pod running in the cluster. This extension-apiserver will normally need to be paired with one or more controllers if active management of the added resources is needed. As a result, the apiserver-builder will actually provide a skeleton for both. For example, when the service-catalog is installed, it provides both the extension-apiserver and controller for the services it provides.
```
Comment (cheftako) [inline docs/concepts/api-extension/apiserver-aggregation.md]: As another example, when the service catalog

File: docs/tasks/access-kubernetes-api/configure-aggregation-layer.md (hunk 17 above)
Comment (cheftako) [inline docs/tasks/access-kubernetes-api/configure-aggregation-layer.md]: It does mean kube-apiserver.

File: docs/concepts/api-extension/apiserver-aggregation.md (hunk 25 above)
Comment (chenopis) [inline docs/concepts/api-extension/apiserver-aggregation.md]: got it

File: docs/tasks/access-kubernetes-api/configure-aggregation-layer.md (hunk 17 above)
Comment (chenopis) [inline docs/tasks/access-kubernetes-api/configure-aggregation-layer.md]: Thanks for the clarification.
